### PR TITLE
GH-521 Adds N-D Data to Column Chart

### PR DESCRIPTION
**Pull request by anmoljagetia**
Fixes GH-521

Signed-off-by: Anmol Jagetia <anmoljagetia@gmail.com>

### CODE REVIEW

File: src/chart/Column.js [hunk 1]
```diff
@@ -1,83 +1,93 @@
...
-        root.chart_Column = factory(root.d3, root.chart_XYAxis, root.api_I2DChart);
+        root.chart_Column = factory(root.d3, root.chart_XYAxis, root.api_INDChart);
     }
-}(this, function (d3, XYAxis, I2DChart) {
+}(this, function (d3, XYAxis, INDChart) {
     function Column(target) {
         XYAxis.call(this);
-        I2DChart.call(this);
-
-        this._linearGap = 25;
```
Comment (GordonSmith) [inline src/chart/Column.js]: Why deleting _linearGap?

File: src/chart/Column.js [hunk 2]
```diff
@@ -1,83 +1,93 @@
...
+        INDChart.call(this);
     }
     Column.prototype = Object.create(XYAxis.prototype);
     Column.prototype._class += " chart_Column";
-    Column.prototype.implements(I2DChart.prototype);
+    Column.prototype.implements(INDChart.prototype);
 
     Column.prototype.publish("paletteID", "default", "set", "Palette ID", Column.prototype._palette.switch(),{tags:['Basic','Shared']});
 
+    Column.prototype.testData2 = function () {
```
Comment (GordonSmith) [inline src/chart/Column.js]: Where is this from?

File: src/chart/Column.js (hunk 1 above)
Comment (anmoljagetia) [inline src/chart/Column.js]: I think I removed it by mistake, during the merge. Will fix this.

File: src/chart/Column.js [hunk 3]
```diff
@@ -1,83 +1,93 @@
...
+                title
+                    .text(function (d, idx) { return dataRow[0] + " (" + d + "," + " " + context._columns[idx + 1] + ")"; })
+                ;
+
+                columnRect.exit().transition()
+                    .remove()
+                ;
+        });
 
-        column.exit().transition()
```
Comment (GordonSmith) [inline src/chart/Column.js]: This is still needed,

File: src/chart/Column.js (hunk 2 above)
Comment (anmoljagetia) [inline src/chart/Column.js]: Again, mistake from the merge, this has been moved into the XYAxis now (I think!)

File: src/chart/Column.js [hunk 4]
```diff
@@ -23,61 +23,80 @@
...
-        var column = this.svgData.selectAll(".columnRect")
+        var dataLen = 10;
+        switch (this.xAxisType()) {
+            case "ordinal":
+              dataLen = this.dataScale.rangeBand();
+               break;
+            case "linear":
+            case "time":
+               dataLen = Math.max(Math.abs(this.dataScale(2) - this.dataScale(1)) * (100 - this._linearGap) / 100, 1);
+               // dataPos -= dataLen / 2;
```
Comment (GordonSmith) [inline src/chart/Column.js]: We must work this back into your xpos for the linear / time xAxisType

File: src/chart/Column.js [hunk 5]
```diff
@@ -23,61 +23,91 @@
...
 
-        var column = this.svgData.selectAll(".columnRect")
+        var dataLen = 10;
+        switch (this.xAxisType()) {
+            case "ordinal":
+              dataLen = this.dataScale.rangeBand();
+               break;
+            case "linear":
+            case "time":
+               dataLen = Math.max(Math.abs(this.dataScale(2) - this.dataScale(1)) * (100 - this._linearGap) / 100, 1);
```
Comment (GordonSmith) [inline src/chart/Column.js]: Instead of ```, 1);``` make it ```, dataLen);```

File: src/chart/Column.js [hunk 6]
```diff
@@ -23,61 +23,91 @@
 
         this._palette = this._palette.switch(this.paletteID());
 
-        var column = this.svgData.selectAll(".columnRect")
+        var dataLen = 10;
+        switch (this.xAxisType()) {
+            case "ordinal":
+              dataLen = this.dataScale.rangeBand();
```
Comment (GordonSmith) [inline src/chart/Column.js]: indentation

File: src/chart/Column.js [hunk 7]
```diff
@@ -23,61 +23,91 @@
...
-        column.transition()
-            .style("fill", function (d) { return context._palette(d[0]); })
-            .each(function (d) {
+        column
+            .each(function (dataRow, i) {
+                var dataPos = context.dataScale(dataRow[0]);
                 var element = d3.select(this);
-                var dataPos = context.dataScale(d[0]);
-                var dataLen = 10;
+                var offset = 0;
```
Comment (GordonSmith) [inline src/chart/Column.js]: The offset can be calculated inside the same switch statement that the dataLen is calculated.

File: src/chart/Column.js [hunk 8]
```diff
@@ -23,61 +23,91 @@
...
-                        dataLen = context.dataScale.rangeBand();
+                        offset = 0;
                         break;
                     case "linear":
                     case "time":
-                        dataLen = Math.max(Math.abs(context.dataScale(2) - context.dataScale(1)) * (100 - context._linearGap) / 100, 1);
-                        dataPos -= dataLen / 2;
-                        break;
+                        offset = -dataLen/2;
+                       break;
```
Comment (GordonSmith) [inline src/chart/Column.js]: indentation

File: src/chart/Column.js [hunk 9]
```diff
@@ -23,61 +23,91 @@
...
                     ;
                 } else {
-                    element.transition()
-                        .attr("x", 0)
-                        .attr("width", valuePos)
-                        .attr("y", dataPos)
-                        .attr("height", dataLen)
+                    columnRect.transition()
+                        .attr("class", "columnRect")
+                        .attr("y", function (d, idx) { return context.dataScale(dataRow[0]) + columnScale(context._columns[idx + 1]);})
```
Comment (GordonSmith) [inline src/chart/Column.js]: This will need the offset also

File: src/chart/Column.js [hunk 10]
```diff
@@ -23,61 +23,82 @@
...
+                break;
+        }
+
+        var columnScale = d3.scale.ordinal()
+            .domain(context._columns.filter(function (d, idx) { return idx > 0; }))
+            .rangeRoundBands([0, dataLen])
+        ;
+
+        var column = this.svgData.selectAll(".dataRow")
+            .data(this.data())
```
Comment (GordonSmith) [inline src/chart/Column.js]: Should be this.formattedData() as per old line 27

**Comment by GordonSmith** [comment]
Looks good.

**Pull request merged**